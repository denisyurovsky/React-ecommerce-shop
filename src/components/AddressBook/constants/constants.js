export const LISTBOX_PADDING = 8;
export const DOUBLE_LISTBOX_PADDING = LISTBOX_PADDING * 2;
export const LIMIT_NUMBER_ITEMS = 8;
export const ITEM_SIZE_LG = 48;
export const ITEM_SIZE_MD = 36;
