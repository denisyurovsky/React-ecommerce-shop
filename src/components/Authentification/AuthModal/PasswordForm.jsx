import { VisibilityOff, Visibility } from '@mui/icons-material';
import { Checkbox, TextField, InputAdornment } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState, useEffect } from 'react';

const PasswordForm = ({
  label,
  isError,
  errorMessage,
  password,
  onChange,
  disabled,
  hidden,
  onHiddenChange,
}) => {
  const [isHidden, setIsHidden] = useState(hidden);
  const [labelFocus, setLabelFocus] = useState(false);

  const onInputPasswrodFocus = () => {
    setLabelFocus(true);
  };

  const onInputPasswordBlur = () => {
    setLabelFocus(false);
  };

  const changeVisibility = () => {
    setIsHidden(!isHidden);
    onHiddenChange();
  };

  useEffect(() => {
    setIsHidden(hidden);
  }, [hidden]);

  return (
    <TextField
      error={isError}
      helperText={isError ? errorMessage : null}
      autoComplete="off"
      fullWidth
      variant="standard"
      margin="normal"
      label={label}
      InputLabelProps={{ shrink: password.length || labelFocus }}
      onFocus={onInputPasswrodFocus}
      onBlur={onInputPasswordBlur}
      value={password}
      onChange={onChange}
      type={isHidden ? 'password' : 'text'}
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <Checkbox
              data-testid="visibility"
              icon={<Visibility />}
              checkedIcon={<VisibilityOff />}
              checked={isHidden}
              onChange={changeVisibility}
              disabled={disabled}
            />
          </InputAdornment>
        ),
      }}
      disabled={disabled}
    />
  );
};

PasswordForm.defaultProps = {
  label: 'Password',
  isError: false,
  errorMessage: null,
  disabled: false,
  hidden: true,
  onHiddenChange: () => {},
};

PasswordForm.propTypes = {
  password: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  isError: PropTypes.bool,
  errorMessage: PropTypes.string,
  disabled: PropTypes.bool,
  hidden: PropTypes.bool,
  onHiddenChange: PropTypes.func,
};

export default PasswordForm;
