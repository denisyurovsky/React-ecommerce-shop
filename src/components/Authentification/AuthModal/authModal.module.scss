@import "../../../assets/styles/breakpoints.scss";

.navigation {
  margin-top: -2rem;
}

.content {
  padding-left: 2rem;
  padding-right: 2rem;
}

@media (max-width: $sm) {
  .content {
    padding: 0;
  }

  .navigation {
    margin-top: -1rem;
  }
}
