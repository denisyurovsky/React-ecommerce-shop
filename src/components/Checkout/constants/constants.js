export const PANEL = {
  DELIVERY_ADDRESS: 'Delivery Address',
  PERSONAL_INFORMATION: 'Personal Information',
  PERSONAL_ADDRESS: 'Personal Address',
  PAYMENT_METHOD: 'Payment method',
  ORDER_CONFIRMATION: 'Order confirmation',
};
