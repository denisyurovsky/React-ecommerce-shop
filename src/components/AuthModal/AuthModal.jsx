import CloseIcon from '@mui/icons-material/Close';
import { Tabs, Tab, Box, Modal, IconButton, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { authStatus } from '../../helpers/constants/authConstants';

import Registration from './Registration/Registration';
import SignIn from './SignIn/SignIn';

import styles from './authModal.module.scss';

const { FULFILLED, IDLE, LOADING, FAILED, LOCKED } = authStatus;

const AuthModal = ({
  registerUser,
  registerStatus,
  loginUser,
  loginStatus,
  errorMessage,
  resetError,
}) => {
  const [value, setValue] = useState('old');
  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (event, newValue) => setValue(newValue);
  const handleClose = () => setIsOpen(false);
  const handleOpen = () => setIsOpen(true);

  const isLogged = loginStatus === FULFILLED || registerStatus === FULFILLED;

  const open =
    isOpen && loginStatus !== FULFILLED && registerStatus !== FULFILLED;

  return (
    <>
      {isLogged ? (
        <Button
          data-testid="btn-profile"
          className={styles.button}
          variant="outlined"
        >
          Profile
        </Button>
      ) : (
        <Button
          data-testid="btn-login"
          onClick={handleOpen}
          className={styles.button}
          variant="outlined"
        >
          Sign in
        </Button>
      )}
      <Modal open={open} onClose={handleClose} className={styles.container}>
        <Box className={styles.modal}>
          <Box>
            <Tabs
              centered
              value={value}
              onChange={handleChange}
              textColor="primary"
              indicatorColor="primary"
            >
              <Tab
                label="Sign in"
                value="old"
                sx={{ width: 0.25, minWidth: 100 }}
                onClick={resetError}
              />
              <Tab
                label="Sign up"
                value="new"
                sx={{ width: 0.25, minWidth: 100 }}
                onClick={resetError}
              />
            </Tabs>
            <IconButton
              className={styles.cross}
              color="primary"
              onClick={handleClose}
              size="small"
            >
              <CloseIcon />
            </IconButton>
          </Box>
          <Box className={styles.forms}>
            {value === 'new' ? (
              <Registration
                sendForm={registerUser}
                status={registerStatus}
                errorMessage={errorMessage}
              />
            ) : (
              <SignIn
                sendForm={loginUser}
                status={loginStatus}
                errorMessage={errorMessage}
              />
            )}
          </Box>
        </Box>
      </Modal>
    </>
  );
};

AuthModal.defaultProps = {
  errorMessage: '',
  registerStatus: IDLE,
  loginStatus: IDLE,
};

AuthModal.propTypes = {
  resetError: PropTypes.func.isRequired,
  registerUser: PropTypes.func.isRequired,
  loginUser: PropTypes.func.isRequired,
  registerStatus: PropTypes.oneOf([IDLE, LOADING, FULFILLED, FAILED]),
  loginStatus: PropTypes.oneOf([IDLE, LOADING, FULFILLED, FAILED, LOCKED]),
  errorMessage: PropTypes.string,
};

export default AuthModal;
