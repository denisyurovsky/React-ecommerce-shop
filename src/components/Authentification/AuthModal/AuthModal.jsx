import CloseIcon from '@mui/icons-material/Close';
import { Tabs, Tab, Box, Modal, IconButton, Button } from '@mui/material';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { authStatus } from '../../../helpers/constants/authConstants';

import Registration from './Registration/Registration';
import SignIn from './SignIn/SignIn';

import styles from './authModal.module.scss';

const { FULFILLED, IDLE, PENDING, REJECTED, LOCKED } = authStatus;

const AuthModal = ({
  registerUser,
  registerStatus,
  loginUser,
  loginStatus,
  loginErrorMessage,
  registerErrorMessage,
  resetError,
}) => {
  const [value, setValue] = useState('old');
  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (event, newValue) => {
    resetError();
    setValue(newValue);
  };
  const handleClose = () => {
    resetError();
    setIsOpen(false);
  };
  const handleOpen = () => setIsOpen(true);

  const open =
    isOpen && loginStatus !== FULFILLED && registerStatus !== FULFILLED;

  return (
    <>
      <Button
        data-testid="btn-login"
        onClick={handleOpen}
        variant="outlined"
        color="secondary"
      >
        Sign in
      </Button>
      <Modal open={open} onClose={handleClose} className={styles.container}>
        <Box className={styles.modal}>
          <Box>
            <Tabs
              centered
              value={value}
              onChange={handleChange}
              textColor="primary"
              indicatorColor="primary"
            >
              <Tab
                label="Sign in"
                value="old"
                sx={{ width: 0.25, minWidth: 100 }}
              />
              <Tab
                label="Sign up"
                value="new"
                sx={{ width: 0.25, minWidth: 100 }}
              />
            </Tabs>
            <IconButton
              className={styles.cross}
              color="primary"
              onClick={handleClose}
              size="small"
            >
              <CloseIcon />
            </IconButton>
          </Box>
          <Box className={styles.forms}>
            {value === 'new' ? (
              <Registration
                sendForm={registerUser}
                status={registerStatus}
                errorMessage={registerErrorMessage}
              />
            ) : (
              <SignIn
                sendForm={loginUser}
                status={loginStatus}
                errorMessage={loginErrorMessage}
              />
            )}
          </Box>
        </Box>
      </Modal>
    </>
  );
};

AuthModal.defaultProps = {
  registerErrorMessage: '',
  loginErrorMessage: '',
  registerStatus: IDLE,
  loginStatus: IDLE,
};

AuthModal.propTypes = {
  resetError: PropTypes.func.isRequired,
  registerUser: PropTypes.func.isRequired,
  loginUser: PropTypes.func.isRequired,
  registerStatus: PropTypes.oneOf([IDLE, PENDING, FULFILLED, REJECTED]),
  loginStatus: PropTypes.oneOf([IDLE, PENDING, FULFILLED, REJECTED, LOCKED]),
  loginErrorMessage: PropTypes.string,
  registerErrorMessage: PropTypes.string,
};

export default AuthModal;
