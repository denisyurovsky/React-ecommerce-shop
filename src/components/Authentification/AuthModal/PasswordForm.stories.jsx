import React from 'react';

import PasswordForm from './PasswordForm';

export default {
  title: 'AuthModal/PasswordField',
  component: PasswordForm,
  argTypes: { onChange: { action: 'Typing...' } },
};

const Template = (args) => <PasswordForm {...args} />;

export const Default = Template.bind({});
Default.args = {
  password: 'password',
  isError: false,
  disabled: false,
  errorMessage: 'Incorrect login or password',
};

export const Error = Template.bind({});
Error.args = {
  password: 'wrongpassword',
  isError: true,
  disabled: false,
  errorMessage: 'Incorrect login or password',
};

export const Disabled = Template.bind({});
Disabled.args = {
  password: 'password',
  isError: false,
  disabled: true,
  errorMessage: 'Incorrect login or password',
};
