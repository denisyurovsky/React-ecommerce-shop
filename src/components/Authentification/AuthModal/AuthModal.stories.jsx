import React from 'react';

import AuthModal from './AuthModal';

export default {
  title: 'AuthModal/Modal',
  component: AuthModal,
  argTypes: {
    registerUser: { action: 'Sending post request' },
    loginUser: { action: 'Sending post request' },
  },
};

const Template = (args) => <AuthModal {...args} />;

export const Default = Template.bind({});
Default.args = {
  registerStatus: 'idle',
  loginStatus: 'idle',
  errorMessage: 'Incorrect password or login',
};

export const Loading = Template.bind({});
Loading.storyName = 'Loading login';
Loading.args = {
  registerStatus: 'idle',
  loginStatus: 'loading',
  errorMessage: 'Incorrect password or login',
};

export const Error = Template.bind({});
Error.storyName = 'Error in registration';
Error.args = {
  registerStatus: 'failed',
  loginStatus: 'idle',
  errorMessage: 'Incorrect password or login',
};

export const Succeed = Template.bind({});
Succeed.storyName = 'Successful login';
Succeed.args = {
  registerStatus: 'idle',
  loginStatus: 'succeeded',
  errorMessage: 'Incorrect password or login',
};

export const Locked = Template.bind({});
Locked.storyName = 'Locked login';
Locked.args = {
  registerStatus: 'idle',
  loginStatus: 'locked',
  errorMessage: 'Many wrong attempts, try again later',
};
