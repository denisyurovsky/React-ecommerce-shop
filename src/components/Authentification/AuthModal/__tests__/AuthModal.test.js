import { render, fireEvent, screen } from '@testing-library/react';
import React from 'react';

import { authStatus } from '../../../../constants/authConstants';
import AuthModal from '../AuthModal';

const { FULFILLED } = authStatus;

const openModal = () => {
  const modalBtn = screen.getByTestId('btn-login');

  fireEvent.click(modalBtn);
};

describe('AuthModal snapshot', () => {
  it('should render a valid snapshot', () => {
    const { asFragment } = render(
      <AuthModal
        registerUser={() => {}}
        loginUser={() => {}}
        resetError={() => {}}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Navigation functionality tests', () => {
  beforeEach(() => {
    render(
      <AuthModal
        registerUser={() => {}}
        loginUser={() => {}}
        resetError={() => {}}
      />
    );
    openModal();
  });

  it('should open a SignUp forms', () => {
    const navLoginBtn = screen.getByRole('tab', { selected: true });

    expect(navLoginBtn).toHaveAttribute('aria-selected', 'true');
  });

  it('should switch between SignUp and SigIn', () => {
    const navRegisterBtn = screen.getByRole('tab', { selected: false });
    const navLoginBtn = screen.getByRole('tab', { selected: true });

    expect(navLoginBtn).toHaveAttribute('aria-selected', 'true');
    expect(navRegisterBtn).toHaveAttribute('aria-selected', 'false');

    fireEvent.click(navRegisterBtn);
    expect(navLoginBtn).toHaveAttribute('aria-selected', 'false');
    expect(navRegisterBtn).toHaveAttribute('aria-selected', 'true');
  });
});

describe('Closing functionality', () => {
  it('should close after closeIcon is pressed', () => {
    const { getByTestId, queryByRole } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        resetError={() => {}}
      />
    );

    openModal();
    const cross = getByTestId('CloseIcon');

    fireEvent.click(cross);

    expect(queryByRole('presentation')).toBeNull();
  });

  it('should close after successful login', () => {
    const { queryByRole } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        loginStatus={FULFILLED}
        resetError={() => {}}
      />
    );

    expect(queryByRole('presentation')).toBeNull();
  });

  it('should close after successful registration', () => {
    const { queryByRole } = render(
      <AuthModal
        loginUser={() => {}}
        registerUser={() => {}}
        registerStatus={FULFILLED}
        resetError={() => {}}
      />
    );

    expect(queryByRole('presentation')).toBeNull();
  });
});
