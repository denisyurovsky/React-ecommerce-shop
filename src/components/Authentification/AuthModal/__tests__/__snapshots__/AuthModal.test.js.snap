// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal snapshot should render a valid snapshot 1`] = `
<DocumentFragment>
  <button
    class="MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-wnpvqa-MuiButtonBase-root-MuiButton-root"
    data-testid="btn-login"
    tabindex="0"
    type="button"
  >
    Sign in
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
